### DISCUSSION TITLE
How can we define the range for one of the input parameters while generating API tests

### DISCUSSION THREAD

**Discussion by soumyadipmitralpl** [Cornutum/tcases#198] [Q&A] [answered]
Hi Kerry,

While generating API tests from an openAPI spec, can you kindly show us an example of how the input xml file will look like, where we will define the range of values that one of the input parameters can have?

**Comment by kerrykimbrough** [comment, marked as answer]
All of the information about input parameters must come from the OpenAPI definition. For example, if you have an input parameter that must be an integer between 0 and 100, you can define its value range like this:

```json
...
"parameters": [
    {
        "name": "myParam",
        "schema": {
            "type": "integer",
            "minimum": 0,
            "maximum": 100
        }
    }
]
...
```

**Reply by AnirudhChaudhary** [reply]
To follow up on this, could you explain what the section "Modeling the Input Space" means in relation to this topic. Based on my understanding, it seems like we can create an xml file called {Project-Name}-Input.xml that would provide defined inputs to TCases.
Does this xml file define parameters for the functions that are created in the actual test case document or does it define parameters for the paths/requests in the OpenAPI spec.

**Reply by kerrykimbrough** [reply]
*Tcases: The Complete Guide* is a guide to the original Tcases tool. Tcases for OpenAPI uses the same underlying technology, but it is a different tool that is used to handle OpenAPI definitions.

TFOA uses the OpenAPI definition as the input space model -- no XML is involved. More precisely, TFOA converts the OpenAPI definition into a Tcases input model, then uses Tcases to generate (abstract) test cases, then finally converts the abstract test cases into actual test code.

*The Complete Guide* is good background information, if you want to understand more about how the basic Tcases tool works. But not required if you only use TFOA.

**Reply by AnirudhChaudhary** [reply]
Thank you for that clarification. This now means that if we want to use TCases for OpenAPI, in order to have control over the input parameters for the variables in the paths, we will have to manually put that into the OpenAPI spec to begin with.

**Reply by AnirudhChaudhary** [reply]
Also do these input parameters only work for the test cases that TCases is testing that specific parameter? I am seeing that for test cases that involve some other variable, the constraints for the variable are not being applied.

for example I define a variable "x" to be between 1 and 5 but I see that in a different test that is testing "y", x is not bounded between 1 and 5. Could you clarify?

**Reply by kerrykimbrough** [reply]
> for example I define a variable "x" to be between 1 and 5 but I see that in a different test that is testing "y", x is not bounded between 1 and 5. 

If the OpenAPI definition declares that X has minimum=1 and maximum=5, the only test case in which X lies outside that range should be a failure case expecting a BadRequest response.

If that is not what you are seeing, then please submit an issue with a sample OpenAPI definition that will reproduce this problem.
